File: src/main/java/io/jjong/algorithm/word/WordCounterUniCode.java
package io.jjong.algorithm.word;

import java.util.HashMap;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * create on 2023/01/02. create by IntelliJ IDEA.
 *
 * <p> 클래스 설명 </p>
 *
 * @author Jongsang Han(henry)
 * @version 1.0
 * @see
 * @since 1.0
 */
public class WordCounterUniCode {

  public static void main(String[] args) {
    String str = "aaabbbccceeefff12345👍🔥國小心♥♥♥♥♥";

    System.out.println(countDuplicateCharacters(str));
    System.out.println(countDuplicateCharacters2(str));
  }

  public static Map<String, Integer> countDuplicateCharacters(String str) {
    // 맵을 이용하여 문자를 키, 빈도수를 값으로 저장 한다.
    Map<String, Integer> result = new HashMap<>();

    for (int i=0; i <str.length(); i++) {
      int cp = str.codePointAt(i); // index에 해당하는 문자를 숫자로 변환한 값을 리턴해주는 메소드이다.
      String ch = String.valueOf(Character.toChars(cp));
      System.out.println("ch : " + ch + ", cp : " + cp +  ", charCount : " + Character.charCount(cp));
      if (Character.charCount(cp) == 2) { // 2는 대리 쌍을 뜻 함.
        i++;
      }

      result.compute(ch, (k, v) -> (v == null) ? 1 : ++v);
    }
    return result;

  }

  public static Map<String, Long> countDuplicateCharacters2(String str) {
    return str.codePoints()
        .mapToObj(c -> String.valueOf(Character.toChars(c)))
        .collect(Collectors.groupingBy( // 문자를 분류 하고
            c -> c, Collectors.counting()) // 카운팅
        );
  }

}

File: src/main/kotlin/io/jjong/algorithm/word/WordCounterUniCode.kt
package io.jjong.algorithm.word

/**
 * 유니코드 코드포인트 단위로 각 문자의 등장 횟수를 센다.
 * 이모지·서로게이트 쌍(charCount == 2)을 한 글자로 올바르게 집계한다.
 */
object WordCounterUniCode {
    /** 코드포인트를 직접 순회하며 서로게이트 쌍을 건너뛰어 집계한다. */
    fun countDuplicateCharacters(str: String): Map<String, Int> {
        val result = HashMap<String, Int>()
        var i = 0
        while (i < str.length) {
            val cp = str.codePointAt(i) // 해당 위치의 문자를 코드포인트(정수)로
            val ch = String(Character.toChars(cp))
            println("ch : $ch, cp : $cp, charCount : ${Character.charCount(cp)}")
            if (Character.charCount(cp) == 2) { // 2는 대리 쌍(surrogate pair)을 뜻한다.
                i++
            }
            result[ch] = (result[ch] ?: 0) + 1
            i++
        }
        return result
    }

    /** 코드포인트 스트림을 grouping으로 집계한다. */
    fun countDuplicateCharacters2(str: String): Map<String, Long> {
        return str.codePoints().toArray()
            .map { String(Character.toChars(it)) }
            .groupingBy { it }
            .eachCount()
            .mapValues { (_, count) -> count.toLong() }
    }
}

fun main() {
    val str = "aaabbbccceeefff12345👍🔥國小心♥♥♥♥♥"

    println(WordCounterUniCode.countDuplicateCharacters(str))
    println(WordCounterUniCode.countDuplicateCharacters2(str))
}
